python/tests/test_ld_matrix.py

```diff
@@ -699,6 +735,210 @@ def r2_summary_func(
+    """
+    sample_set_sizes = params["sample_set_sizes"]
+    for k in range(state_dim):
+        n = sample_set_sizes[k]
+        p_AB = state[0, k] / float(n)
+        p_Ab = state[1, k] / float(n)
+        p_aB = state[2, k] / float(n)
+
+        p_A = p_AB + p_Ab
+        p_B = p_AB + p_aB
+
+        result[k] = p_AB - (p_A * p_B)
+
+
+def D2_summary_func(
+    state_dim: int, state: np.ndarray, result: np.ndarray, params: Dict[str, Any]
+) -> None:
+    sample_set_sizes = params["sample_set_sizes"]
+    for k in range(state_dim):
+        n = sample_set_sizes[k]
+        p_AB = state[0, k] / float(n)
+        p_Ab = state[1, k] / float(n)
+        p_aB = state[2, k] / float(n)
+
+        p_A = p_AB + p_Ab
+        p_B = p_AB + p_aB
+
+        result[k] = p_AB - (p_A * p_B)
+        result[k] = result[k] * result[k]
+
+
+def D_prime_summary_func(
+    state_dim: int, state: np.ndarray, result: np.ndarray, params: Dict[str, Any]
+) -> None:
+    """Summary function for the D_prime statistic. We first compute the proportion of
+    AB, A, and B haplotypes, then we compute the D_prime statistic, storing the outputs
+    in the result vector, one entry per sample set.
+
+    :param state_dim: Number of sample sets.
+    :param state: Counts of 3 haplotype configurations for each sample set.
+    :param result: Vector of length state_dim to store the results in.
+    :param params: Parameters for the summary function.
+    """
+    sample_set_sizes = params["sample_set_sizes"]
+    for k in range(state_dim):
+        n = sample_set_sizes[k]
+        p_AB = state[0, k] / float(n)
+        p_Ab = state[1, k] / float(n)
+        p_aB = state[2, k] / float(n)
+
+        p_A = p_AB + p_Ab
+        p_B = p_AB + p_aB
+
+        D = p_AB - (p_A * p_B)
+        if D == 0:
+            result[k] = 0
+        elif D > 0:
+            result[k] = D / min(p_A * (1 - p_B), (1 - p_A) * p_B)
+        else:
+            result[k] = D / min(p_A * p_B, (1 - p_A) * (1 - p_B))
+
+
+def r_summary_func(
+    state_dim: int, state: np.ndarray, result: np.ndarray, params: Dict[str, Any]
+) -> None:
+    """Summary function for the r statistic. We first compute the proportion of
+    AB, A, and B haplotypes, then we compute the r statistic, storing the outputs
+    in the result vector, one entry per sample set.
+
+    :param state_dim: Number of sample sets.
+    :param state: Counts of 3 haplotype configurations for each sample set.
+    :param result: Vector of length state_dim to store the results in.
+    :param params: Parameters for the summary function.
+    """
+    sample_set_sizes = params["sample_set_sizes"]
+    for k in range(state_dim):
+        n = sample_set_sizes[k]
+        p_AB = state[0, k] / n
+        p_Ab = state[1, k] / n
+        p_aB = state[2, k] / n
+
+        p_A = p_AB + p_Ab
+        p_B = p_AB + p_aB
+
+        D = p_AB - (p_A * p_B)
+        denom = p_A * p_B * (1 - p_A) * (1 - p_B)
+
+        if denom == 0 and D == 0:
```

Comment:
See point above about NaN results. We found for the other stats API that it was easier to just let the division by zero happen and return NaN where the stat is undefined. Using a capturewarnings stops annoying warnings happening.